org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend

```diff
@@ -411,11 +433,33 @@ import org.eclipse.xtext.validation.Issue
 	}
 
 	override codeLens(CodeLensParams params) {
-		throw new UnsupportedOperationException("TODO: auto-generated method stub")
+		return requestManager.runRead[ cancelIndicator |
+			val uri = params.textDocument.uri.toUri
+			val resourceServiceProvider = uri.resourceServiceProvider
+			val codeLensService = resourceServiceProvider?.get(ICodeLensService)
+			if (codeLensService === null)
+				return emptyList
+
+			return workspaceManager.doRead(uri) [ document, resource |
+				codeLensService.computeCodeLenses(document, resource, params, cancelIndicator)
+			]
+		]
 	}
 
 	override resolveCodeLens(CodeLens unresolved) {
-		return CompletableFuture.completedFuture(unresolved)
+		return CompletableFutures.computeAsync[ cancelChecker |
+			var CodeLens current = unresolved
+			for (language : this.allLanguages) {
+				val codeLensService = language.get(ICodeLensResolver)
+				if (codeLensService !== null) {
+					current = codeLensService.resolveCodeLens(this.access, current, [ 
```

Comment:
Should not we pass a special access that cancels all read requests if the user cancels this future? It looks like the user will end up with 2 cancel indicators at some point and should check both all the time, otherwise, he will miss one or another cancellation.